File: src/border.zig
// The borders below are adapted from Nushell table themes, which in turn are
// built on the `tabled` crate's style system. THANKS GUYS
//
// Omitted on purpose:
// - `default`, which is just another name for `rounded`
//

// Each "specimen" is a tiny canonical table, which we parse below
const specimens = struct {
    const ascii_rounded =
        \\.-----.
        \\|A|B|C|
        \\|D|E|F|
        \\|G|H|I|
        \\'-----'
    ;
    const basic =
        \\+-+-+-+
        \\|A|B|C|
        \\+-+-+-+
        \\|D|E|F|
        \\+-+-+-+
        \\|G|H|I|
        \\+-+-+-+
    ;
    const basic_compact =
        \\+-+-+-+
        \\|A|B|C|
        \\|D|E|F|
        \\|G|H|I|
        \\+-+-+-+
    ;
    const compact =
        \\─┬─┬─
        \\A│B│C
        \\─┼─┼─
        \\D│E│F
        \\G│H│I
        \\─┴─┴─
    ;
    const compact_double =
        \\═╦═╦═
        \\A║B║C
        \\═╬═╬═
        \\D║E║F
        \\G║H║I
        \\═╩═╩═
    ;
    const dots =
        \\.......
        \\:A:B:C:
        \\:D:E:F:
        \\:G:H:I:
        \\:.:.:.:
    ;
    const double =
        \\╔═╦═╦═╗
        \\║A║B║C║
        \\╠═╬═╬═╣
        \\║D║E║F║
        \\║G║H║I║
        \\╚═╩═╩═╝
    ;
    const heavy =
        \\┏━┳━┳━┓
        \\┃A┃B┃C┃
        \\┣━╋━╋━┫
        \\┃D┃E┃F┃
        \\┃G┃H┃I┃
        \\┗━┻━┻━┛
    ;
    const light =
        \\A B C
        \\─────
        \\D E F
        \\G H I
    ;
    const markdown =
        \\|A|B|C|
        \\|-|-|-|
        \\|D|E|F|
        \\|G|H|I|
    ;
    const none =
        \\A B C
        \\D E F
        \\G H I
    ;
    const psql =
        \\A|B|C
        \\-+-+-
        \\D|E|F
        \\G|H|I
    ;
    const reinforced =
        \\┏─┬─┬─┓
        \\│A│B│C│
        \\│D│E│F│
        \\│G│H│I│
        \\┗─┴─┴─┛
    ;
    const restructured =
        \\= = =
        \\A B C
        \\= = =
        \\D E F
        \\G H I
        \\= = =
    ;

    // note: default
    const rounded =
        \\╭─┬─┬─╮
        \\│A│B│C│
        \\├─┼─┼─┤
        \\│D│E│F│
        \\│G│H│I│
        \\╰─┴─┴─╯
    ;
    const single =
        \\┌─┬─┬─┐
        \\│A│B│C│
        \\├─┼─┼─┤
        \\│D│E│F│
        \\│G│H│I│
        \\└─┴─┴─┘
    ;
    const thin =
        \\┌─┬─┬─┐
        \\│A│B│C│
        \\├─┼─┼─┤
        \\│D│E│F│
        \\├─┼─┼─┤
        \\│G│H│I│
        \\└─┴─┴─┘
    ;
    const with_love =
        \\❤❤❤❤❤
        \\A❤B❤C
        \\❤❤❤❤❤
        \\D❤E❤F
        \\G❤H❤I
        \\❤❤❤❤❤
    ;
};

// Main parsed border type consumed by render.
//
//                              ↙ top
//   ┌────────┬────────┬──────┐
//   │ fruit  │ color  │  qt  │
//   ├────────┼────────┼──────┤ <- header
//   │ apple  │ red    │  12  │
//   ├────────┼────────┼──────┤ <- content row
//   │ pear   │ green  │  33  │
//   ├────────┼────────┼──────┤ ...
//   ↑        ↑        ↑      ↑
// left      mid      mid   right
//   ..........................
//   ├────────┼────────┼──────┤ <- content row
//   │ plum   │ purple │  27  │
//   └────────┴────────┴──────┘
//                              ↖ bottom
//

pub const Border = struct {
    top: BorderRule,
    header: BorderRule,
    row: BorderRule,
    bottom: BorderRule,
    left: []const u8,
    mid: []const u8,
    right: []const u8,
};

// Content-row separators inferred from the specimen body lines.
const RowStyle = struct {
    left: []const u8,
    mid: []const u8,
    right: []const u8,
};

// Border names for CLI
pub const BorderName = enum {
    ascii_rounded,
    basic,
    basic_compact,
    compact,
    compact_double,
    dots,
    double,
    heavy,
    light,
    markdown,
    none,
    psql,
    reinforced,
    restructured,
    rounded,
    single,
    thin,
    with_love,
};

// Specimens indexed in `BorderName` declaration order.
const specimen_table = [_][]const u8{
    specimens.ascii_rounded,
    specimens.basic,
    specimens.basic_compact,
    specimens.compact,
    specimens.compact_double,
    specimens.dots,
    specimens.double,
    specimens.heavy,
    specimens.light,
    specimens.markdown,
    specimens.none,
    specimens.psql,
    specimens.reinforced,
    specimens.restructured,
    specimens.rounded,
    specimens.single,
    specimens.thin,
    specimens.with_love,
};

// Parsed horizontal separator line.
pub const BorderRule = union(enum) {
    none,

    // Continuous horizontal rule like `────` or `.....`.
    continuous: struct {
        left: []const u8,
        fill: []const u8,
        right: []const u8,
    },

    // Segmented horizontal rule like `├─┼─┤`.
    segmented: struct {
        left: []const u8,
        fill: []const u8,
        mid: []const u8,
        right: []const u8,
    },
};

// UTF-8 glyph span within a specimen line.
const Glyph = struct {
    start: usize,
    end: usize,
};

// Split specimen lines with a fixed small maximum.
const Specimen = struct {
    items: [7][]const u8,
    len: usize,
};

// Convert a named border into the parsed border representation used by render.
pub fn getBorder(border: BorderName) Border {
    return parseSpecimen(specimen_table[@intFromEnum(border)]);
}

// Parse a canonical specimen into the normalized border type.
fn parseSpecimen(input: []const u8) Border {
    const lines = splitLines(input);
    const header_index = findLine(lines, 'A');
    const first_row_index = findLine(lines, 'D');
    const second_row_index = findLine(lines, 'G');
    const row_style = parseRow(lines.items[header_index]);

    return .{
        .top = if (header_index > 0) parseRule(lines.items[0], row_style) else .none,
        .header = if (first_row_index > header_index + 1) parseRule(lines.items[header_index + 1], row_style) else .none,
        .row = if (second_row_index > first_row_index + 1) parseRule(lines.items[first_row_index + 1], row_style) else .none,
        .bottom = if (second_row_index + 1 < lines.len) parseRule(lines.items[second_row_index + 1], row_style) else .none,
        .left = row_style.left,
        .mid = row_style.mid,
        .right = row_style.right,
    };
}

// Parse the content row to infer the left, middle, and right separators.
fn parseRow(line: []const u8) RowStyle {
    const a = std.mem.indexOfScalar(u8, line, 'A').?;
    const b = std.mem.indexOfScalar(u8, line, 'B').?;
    const c = std.mem.indexOfScalar(u8, line, 'C').?;
    const mid = line[a + 1 .. b];

    std.debug.assert(std.mem.eql(u8, mid, line[b + 1 .. c]));

    return .{
        .left = line[0..a],
        .mid = mid,
        .right = line[c + 1 ..],
    };
}

// Parse a horizontal rule by using the inferred row separators as the guide.
// If the middle slots match the fill char, this is a continuous rule like
// `─────` or `.-----.`; otherwise it is segmented like `├─┼─┼─┤`.
fn parseRule(line: []const u8, row_style: RowStyle) BorderRule {
    const left_glyphs = glyphCount(row_style.left);
    const mid_glyphs = glyphCount(row_style.mid);
    const right_glyphs = glyphCount(row_style.right);

    var glyphs: [32]Glyph = undefined;
    const nglyphs = splitGlyphs(line, &glyphs);
    const needed = left_glyphs + right_glyphs + 3 + 2 * mid_glyphs;
    std.debug.assert(nglyphs == needed);

    const fill1 = glyphRange(line, glyphs[0..nglyphs], left_glyphs, left_glyphs + 1);
    const mid1 = glyphRange(line, glyphs[0..nglyphs], left_glyphs + 1, left_glyphs + 1 + mid_glyphs);
    const fill2 = glyphRange(line, glyphs[0..nglyphs], left_glyphs + 1 + mid_glyphs, left_glyphs + 2 + mid_glyphs);
    const mid2 = glyphRange(line, glyphs[0..nglyphs], left_glyphs + 2 + mid_glyphs, left_glyphs + 2 + 2 * mid_glyphs);
    const fill3 = glyphRange(line, glyphs[0..nglyphs], left_glyphs + 2 + 2 * mid_glyphs, left_glyphs + 3 + 2 * mid_glyphs);

    std.debug.assert(std.mem.eql(u8, fill1, fill2));
    std.debug.assert(std.mem.eql(u8, fill1, fill3));

    if (std.mem.eql(u8, fill1, mid1) and std.mem.eql(u8, fill1, mid2)) {
        return .{ .continuous = .{
            .left = glyphRange(line, glyphs[0..nglyphs], 0, left_glyphs),
            .fill = fill1,
            .right = glyphRange(line, glyphs[0..nglyphs], nglyphs - right_glyphs, nglyphs),
        } };
    }

    std.debug.assert(std.mem.eql(u8, mid1, mid2));
    return .{ .segmented = .{
        .left = glyphRange(line, glyphs[0..nglyphs], 0, left_glyphs),
        .fill = fill1,
        .mid = mid1,
        .right = glyphRange(line, glyphs[0..nglyphs], nglyphs - right_glyphs, nglyphs),
    } };
}

// Split the embedded specimen into physical lines.
fn splitLines(input: []const u8) Specimen {
    var out: [7][]const u8 = undefined;
    var it = std.mem.splitScalar(u8, input, '\n');
    var n: usize = 0;
    while (it.next()) |line| : (n += 1) {
        out[n] = line;
    }
    return .{ .items = out, .len = n };
}

// Find the row containing the given specimen marker.
fn findLine(lines: Specimen, marker: u8) usize {
    for (lines.items[0..lines.len], 0..) |line, ii| {
        if (std.mem.indexOfScalar(u8, line, marker) != null) return ii;
    }
    unreachable;
}

// Split a UTF-8 string into glyph spans.
fn splitGlyphs(line: []const u8, out: *[32]Glyph) usize {
    var n: usize = 0;
    var i: usize = 0;
    while (i < line.len) : (n += 1) {
        const len = std.unicode.utf8ByteSequenceLength(line[i]) catch unreachable;
        out[n] = .{ .start = i, .end = i + len };
        i += len;
    }
    return n;
}

// Count UTF-8 glyphs in a line.
fn glyphCount(line: []const u8) usize {
    var n: usize = 0;
    var i: usize = 0;
    while (i < line.len) : (n += 1) {
        i += std.unicode.utf8ByteSequenceLength(line[i]) catch unreachable;
    }
    return n;
}

// Slice a line by glyph positions instead of byte offsets.
fn glyphRange(line: []const u8, glyphs: []const Glyph, start: usize, end: usize) []const u8 {
    if (start == end) return "";
    return line[glyphs[start].start..glyphs[end - 1].end];
}

//
// tests
//

test "all supported borders roundtrip through the specimen parser" {
    const cases = [_]BorderName{
        .ascii_rounded,
        .basic,
        .basic_compact,
        .compact,
        .compact_double,
        .dots,
        .double,
        .heavy,
        .light,
        .markdown,
        .none,
        .psql,
        .reinforced,
        .restructured,
        .rounded,
        .single,
        .thin,
        .with_love,
    };

    for (cases) |border| {
        const want = splitLines(specimen_table[@intFromEnum(border)]);
        const got = getBorder(border);

        var bufs: [7][32]u8 = undefined;
        const lines = [_][]const u8{
            renderRule(&bufs[0], got, got.top),
            renderRow(&bufs[1], got, "A", "B", "C"),
            renderRule(&bufs[2], got, got.header),
            renderRow(&bufs[3], got, "D", "E", "F"),
            renderRule(&bufs[4], got, got.row),
            renderRow(&bufs[5], got, "G", "H", "I"),
            renderRule(&bufs[6], got, got.bottom),
        };
        const have = compactLines(lines);

        try std.testing.expectEqual(want.len, have.len);
        for (want.items[0..want.len], 0..) |line, ii| {
            try std.testing.expectEqualStrings(line, have.items[ii]);
        }
    }
}

test "thin has a row separator but rounded does not" {
    try std.testing.expect(getBorder(.thin).row != .none);
    try std.testing.expect(getBorder(.rounded).row == .none);
}

test "psql omits outer borders" {
    const psql = getBorder(.psql);
    try std.testing.expectEqualStrings("", psql.left);
    try std.testing.expectEqualStrings("", psql.right);
}

test "reinforced has a bottom border but no header separator" {
    const reinforced = getBorder(.reinforced);
    try std.testing.expect(reinforced.header == .none);
    try std.testing.expect(reinforced.bottom != .none);
}

//
// Test-only helpers for specimen roundtrip coverage.
//

fn compactLines(lines: [7][]const u8) Specimen {
    var out: [7][]const u8 = undefined;
    var n: usize = 0;
    for (lines) |line| {
        if (line.len == 0) continue;
        out[n] = line;
        n += 1;
    }
    return .{ .items = out, .len = n };
}

fn renderRule(buf: []u8, style_: Border, line: BorderRule) []const u8 {
    var n: usize = 0;
    switch (line) {
        .none => {},
        .continuous => |rule| {
            append(buf, &n, rule.left);
            for (0..3 + 2 * glyphCount(style_.mid)) |_| append(buf, &n, rule.fill);
            append(buf, &n, rule.right);
        },
        .segmented => |rule| {
            append(buf, &n, rule.left);
            append(buf, &n, rule.fill);
            append(buf, &n, rule.mid);
            append(buf, &n, rule.fill);
            append(buf, &n, rule.mid);
            append(buf, &n, rule.fill);
            append(buf, &n, rule.right);
        },
    }
    return buf[0..n];
}

fn renderRow(buf: []u8, style_: Border, a: []const u8, b: []const u8, c: []const u8) []const u8 {
    var n: usize = 0;
    append(buf, &n, style_.left);
    append(buf, &n, a);
    append(buf, &n, style_.mid);
    append(buf, &n, b);
    append(buf, &n, style_.mid);
    append(buf, &n, c);
    append(buf, &n, style_.right);
    return buf[0..n];
}

fn append(buf: []u8, n: *usize, text: []const u8) void {
    @memcpy(buf[n.* .. n.* + text.len], text);
    n.* += text.len;
}

const std = @import("std");

File: src/border.rs
//! Border definitions.

use crate::{builder::types::Border, util::display_width};

//
// entrypoints
//

/// A fully parsed border definition for rendering
#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) struct BorderDraw {
  pub(crate) top: BorderRule,
  pub(crate) header: BorderRule,
  pub(crate) row: BorderRule,
  pub(crate) bottom: BorderRule,
  pub(crate) left: String,
  pub(crate) mid: String,
  pub(crate) right: String,
}

impl BorderDraw {
  pub(crate) fn chrome_width(&self, ncols: usize) -> usize {
    if ncols == 0 {
      return 0;
    }
    display_width(&self.left)
      + display_width(&self.right)
      + display_width(&self.mid) * ncols.saturating_sub(1)
      + 2 * ncols
  }
}

/// Definition for a "rule", which is a horizontal line
#[derive(Clone, Debug, Eq, PartialEq)]
pub(crate) enum BorderRule {
  None,
  Continuous { left: String, fill: String, right: String },
  Segmented { left: String, fill: String, mid: String, right: String },
}

impl BorderRule {
  pub(crate) fn title_rule(&self, header: &Self) -> Self {
    match (self, header) {
      (BorderRule::Segmented { mid, .. }, BorderRule::Segmented { left, fill, right, .. }) => {
        BorderRule::Segmented { left: left.clone(), fill: fill.clone(), mid: mid.clone(), right: right.clone() }
      }
      _ => header.clone(),
    }
  }

  pub(crate) fn span(&self) -> Self {
    match self {
      BorderRule::None => BorderRule::None,
      BorderRule::Continuous { .. } => self.clone(),
      BorderRule::Segmented { left, fill, right, .. } => {
        BorderRule::Continuous { left: left.clone(), fill: fill.clone(), right: right.clone() }
      }
    }
  }

  pub(crate) fn footer_rule(&self, bottom: &Self) -> Self {
    match (self, bottom) {
      (BorderRule::Segmented { left, fill, right, .. }, BorderRule::Segmented { mid, .. }) => {
        BorderRule::Segmented { left: left.clone(), fill: fill.clone(), mid: mid.clone(), right: right.clone() }
      }
      _ => self.clone(),
    }
  }
}

//
// Specimens use 3×3 cell grids with markers A-I:
//
// 0: ╭─┬─┬─╮  ← top rule
// 1: │A│B│C│  ← header
// 2: ├─┼─┼─┤  ← header separator
// 3: │D│E│F│  ← first data row
// 4: ├─┼─┼─┤  ← row separator
// 5: │G│H│I│  ← second data row
// 6: ╰─┴─┴─╯  ← bottom rule
//

/// Lookup a Border so we can draw it
pub(crate) fn get_border(name: Border) -> BorderDraw {
  #[rustfmt::skip]
  let specimen = match name {
    Border::AsciiRounded  => ".-----.\n|A|B|C|\n|D|E|F|\n|G|H|I|\n'-----'",
    Border::Basic         => "+-+-+-+\n|A|B|C|\n+-+-+-+\n|D|E|F|\n+-+-+-+\n|G|H|I|\n+-+-+-+",
    Border::BasicCompact  => "+-+-+-+\n|A|B|C|\n|D|E|F|\n|G|H|I|\n+-+-+-+",
    Border::Compact       => "─┬─┬─\nA│B│C\n─┼─┼─\nD│E│F\nG│H│I\n─┴─┴─",
    Border::CompactDouble => "═╦═╦═\nA║B║C\n═╬═╬═\nD║E║F\nG║H║I\n═╩═╩═",
    Border::Dots          => ".......\n:A:B:C:\n:D:E:F:\n:G:H:I:\n:.:.:.:",
    Border::Double        => "╔═╦═╦═╗\n║A║B║C║\n╠═╬═╬═╣\n║D║E║F║\n║G║H║I║\n╚═╩═╩═╝",
    Border::Heavy         => "┏━┳━┳━┓\n┃A┃B┃C┃\n┣━╋━╋━┫\n┃D┃E┃F┃\n┃G┃H┃I┃\n┗━┻━┻━┛",
    Border::Light         => "A B C\n─────\nD E F\nG H I",
    Border::Markdown      => "|A|B|C|\n|-|-|-|\n|D|E|F|\n|G|H|I|",
    Border::None          => "A B C\nD E F\nG H I",
    Border::Psql          => "A|B|C\n-+-+-\nD|E|F\nG|H|I",
    Border::Reinforced    => "┏─┬─┬─┓\n│A│B│C│\n│D│E│F│\n│G│H│I│\n┗─┴─┴─┛",
    Border::Restructured  => "= = =\nA B C\n= = =\nD E F\nG H I\n= = =",
    Border::Rounded       => "╭─┬─┬─╮\n│A│B│C│\n├─┼─┼─┤\n│D│E│F│\n│G│H│I│\n╰─┴─┴─╯",
    Border::Single        => "┌─┬─┬─┐\n│A│B│C│\n├─┼─┼─┤\n│D│E│F│\n│G│H│I│\n└─┴─┴─┘",
    Border::Thin          => "┌─┬─┬─┐\n│A│B│C│\n├─┼─┼─┤\n│D│E│F│\n├─┼─┼─┤\n│G│H│I│\n└─┴─┴─┘",
    Border::WithLove      => "❤❤❤❤❤\nA❤B❤C\n❤❤❤❤❤\nD❤E❤F\nG❤H❤I\n❤❤❤❤❤",
  };
  parse_specimen(specimen)
}

//
// internal
//

fn parse_specimen(input: &str) -> BorderDraw {
  let lines = input.lines().collect::<Vec<_>>();

  // find lines in there
  let head = find_line(&lines, 'A');
  let one = find_line(&lines, 'D');
  let two = find_line(&lines, 'G');

  // parse header
  let header = lines[head];
  let a = glyph_index(header, 'A');
  let left = glyph_range(header, 0, a);
  let b = glyph_index(header, 'B');
  let mid = glyph_range(header, a + 1, b);
  let c = glyph_index(header, 'C');
  let right = glyph_range(header, c + 1, header.chars().count());

  BorderDraw {
    top: if head > 0 { parse_rule(lines[0], &left, &mid, &right) } else { BorderRule::None },
    header: if one > head + 1 { parse_rule(lines[head + 1], &left, &mid, &right) } else { BorderRule::None },
    row: if two > one + 1 { parse_rule(lines[one + 1], &left, &mid, &right) } else { BorderRule::None },
    bottom: if two + 1 < lines.len() { parse_rule(lines[two + 1], &left, &mid, &right) } else { BorderRule::None },
    left,
    mid,
    right,
  }
}

fn parse_rule(line: &str, left: &str, mid: &str, right: &str) -> BorderRule {
  let leftg = left.chars().count();
  let fill = glyph_range(line, leftg, leftg + 1);
  let midg = mid.chars().count();

  // positions:  [left][fill][mid][fill][mid]...
  // index:       0    l    l+1  l+1+m ...
  let mid1_start = leftg + 1;
  let mid2_start = leftg + 2 + midg;
  let sep1 = glyph_range(line, mid1_start, mid1_start + midg);
  let sep2 = glyph_range(line, mid2_start, mid2_start + midg);

  let nglyphs = line.chars().count();
  let rightg = right.chars().count();

  let left = glyph_range(line, 0, leftg);
  let right = glyph_range(line, nglyphs - rightg, nglyphs);
  if fill == sep1 && fill == sep2 {
    BorderRule::Continuous { left, fill, right }
  } else {
    BorderRule::Segmented { left, fill, mid: sep1, right }
  }
}

fn find_line(lines: &[&str], marker: char) -> usize {
  lines.iter().position(|line| line.contains(marker)).expect("border specimen contains marker")
}

fn glyph_index(line: &str, marker: char) -> usize {
  line.chars().position(|ch| ch == marker).expect("border specimen contains cell marker")
}

fn glyph_range(line: &str, start: usize, end: usize) -> String {
  line.chars().skip(start).take(end - start).collect()
}

#[cfg(test)]
mod tests {
  use super::*;

  const BORDERS: &[Border] = &[
    Border::AsciiRounded,
    Border::Basic,
    Border::BasicCompact,
    Border::Compact,
    Border::CompactDouble,
    Border::Dots,
    Border::Double,
    Border::Heavy,
    Border::Light,
    Border::Markdown,
    Border::None,
    Border::Psql,
    Border::Reinforced,
    Border::Restructured,
    Border::Rounded,
    Border::Single,
    Border::Thin,
    Border::WithLove,
  ];

  #[test]
  fn test_all_borders_parse() {
    for border in BORDERS {
      let draw = get_border(*border);
      assert_valid_border(*border, &draw);
    }
  }

  #[test]
  fn test_get_border() {
    let basic = get_border(Border::Basic);
    assert_eq!("|", basic.left);
    assert_eq!("|", basic.mid);
    assert_eq!("|", basic.right);

    let rounded = get_border(Border::Rounded);
    assert_eq!("│", rounded.left);
    assert!(matches!(rounded.top, BorderRule::Segmented { .. }));

    let light = get_border(Border::Light);
    assert_eq!("", light.left);
    assert_eq!(" ", light.mid);
    assert_eq!("", light.right);
    assert_eq!(BorderRule::None, light.top);
    assert_eq!(BorderRule::None, light.bottom);

    let thin = get_border(Border::Thin);
    assert!(matches!(thin.header, BorderRule::Segmented { .. }));
    assert!(matches!(thin.row, BorderRule::Segmented { .. }));

    let dots = get_border(Border::Dots);
    assert!(matches!(dots.top, BorderRule::Continuous { .. }));
    assert!(matches!(dots.bottom, BorderRule::Segmented { .. }));
  }

  #[test]
  fn test_span_rule_title_rule_footer_rule() {
    let rounded = get_border(Border::Rounded);
    assert!(matches!(rounded.top.span(), BorderRule::Continuous { .. }));
    assert!(matches!(rounded.top.title_rule(&rounded.header), BorderRule::Segmented { .. }));

    let thin = get_border(Border::Thin);
    assert_eq!(
      BorderRule::Segmented {
        left: "├".to_owned(), fill: "─".to_owned(), mid: "┴".to_owned(), right: "┤".to_owned()
      },
      thin.row.footer_rule(&thin.bottom)
    );
  }

  fn assert_valid_border(border: Border, draw: &BorderDraw) {
    assert_eq!(draw.left.chars().count(), draw.right.chars().count(), "{border:?}");
    assert_valid_rule(border, &draw.top);
    assert_valid_rule(border, &draw.header);
    assert_valid_rule(border, &draw.row);
    assert_valid_rule(border, &draw.bottom);
  }

  fn assert_valid_rule(border: Border, rule: &BorderRule) {
    match rule {
      BorderRule::None => {}
      BorderRule::Continuous { fill, .. } => assert!(!fill.is_empty(), "{border:?}"),
      BorderRule::Segmented { fill, mid, .. } => {
        assert!(!fill.is_empty(), "{border:?}");
        assert!(!mid.is_empty(), "{border:?}");
      }
    }
  }
}
